### ISSUE TITLE
Deepseek_model & test

### ISSUE THREAD

**Issue by ZhigangSu** [ZhigangSu/Zhigang-Su#14]
import torch
import torch.nn as nn
import torch.optim as optim
import pandas as pd
import numpy as np
from torch.utils.data import Dataset, DataLoader, random_split
from collections import Counter
import math
import os
import re

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
EMBED_SIZE = 64
NHEAD = 4
FFN_HIDDEN = 128
NUM_LAYERS = 2
MAX_LENGTH = 20
BATCH_SIZE = 4
EPOCHS = 50
VOCAB_SIZE = 1000
LEARNING_RATE = 0.001
CSV_PATH = 'C:/Users/Ray/Documents/Python/english_spanish_200 (1).csv'

def build_vocab(sentences):
    vocab = {'<pad>':0, '<unk>':1, '<sos>':2, '<eos>':3}
    words = [word for s in sentences for word in s.lower().split()]
    for word, _ in Counter(words).most_common(VOCAB_SIZE - len(vocab)):
        vocab[word] = len(vocab)
    return vocab

class TranslationDataset(Dataset):
    def __init__(self, df, eng_vocab, spa_vocab):
        self.df = df
        self.eng_vocab = eng_vocab
        self.spa_vocab = spa_vocab

    def __len__(self):
        return len(self.df)

    def tokenize(self, text, vocab):
        text = re.sub(r'[^a-záéíóúñü\s]', '', text.lower())
        return [vocab.get(w, vocab['<unk>']) for w in text.split()]

    def __getitem__(self, idx):
        eng = self.tokenize(self.df.iloc[idx]['english'], self.eng_vocab)
        spa = self.tokenize(self.df.iloc[idx]['spanish'], self.spa_vocab)
        src = torch.tensor(eng)
        tgt = torch.tensor([self.spa_vocab['<sos>']] + spa + [self.spa_vocab['<eos>']])
        return src, tgt

def collate_fn(batch):
    src_batch, tgt_batch = zip(*batch)
    src_batch = nn.utils.rnn.pad_sequence(src_batch, padding_value=0, batch_first=True)
    tgt_batch = nn.utils.rnn.pad_sequence(tgt_batch, padding_value=0, batch_first=True)
    return src_batch, tgt_batch

class PositionalEncoding(nn.Module):
    def __init__(self, d_model, max_len=MAX_LENGTH):
        super().__init__()
        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len).unsqueeze(1)
        div = torch.exp(torch.arange(0, d_model, 2) * -(math.log(10000.0) / d_model))
        pe[:, 0::2] = torch.sin(position * div)
        pe[:, 1::2] = torch.cos(position * div)
        self.pe = pe.unsqueeze(0)

    def forward(self, x):
        return x + self.pe[:, :x.size(1)].to(x.device)

class Seq2SeqTransformer(nn.Module):
    def __init__(self, src_vocab_size, tgt_vocab_size):
        super().__init__()
        self.src_embed = nn.Embedding(src_vocab_size, EMBED_SIZE)
        self.tgt_embed = nn.Embedding(tgt_vocab_size, EMBED_SIZE)
        self.pos_enc = PositionalEncoding(EMBED_SIZE)
        self.transformer = nn.Transformer(d_model=EMBED_SIZE, nhead=NHEAD,
                                          num_encoder_layers=NUM_LAYERS,
                                          num_decoder_layers=NUM_LAYERS,
                                          dim_feedforward=FFN_HIDDEN,
                                          batch_first=True)
        self.fc = nn.Linear(EMBED_SIZE, tgt_vocab_size)

    def forward(self, src, tgt, src_mask, tgt_mask):
        src = self.pos_enc(self.src_embed(src))
        tgt = self.pos_enc(self.tgt_embed(tgt))
        output = self.transformer(src, tgt, src_key_padding_mask=src_mask, tgt_mask=tgt_mask)
        return self.fc(output)

def generate_square_subsequent_mask(sz):
    return torch.triu(torch.ones((sz, sz)) * float('-inf'), diagonal=1)

def create_pad_mask(seq, pad_idx=0):
    return (seq == pad_idx)

df = pd.read_csv(CSV_PATH)
eng_vocab = build_vocab(df['english'])
spa_vocab = build_vocab(df['spanish'])

dataset = TranslationDataset(df, eng_vocab, spa_vocab)
train_size = int(0.9 * len(dataset))
train_data, val_data = random_split(dataset, [train_size, len(dataset)-train_size])
train_loader = DataLoader(train_data, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_fn)
val_loader = DataLoader(val_data, batch_size=BATCH_SIZE, collate_fn=collate_fn)

model = Seq2SeqTransformer(len(eng_vocab), len(spa_vocab)).to(DEVICE)
optimizer = optim.Adam(model.parameters(), lr=LEARNING_RATE)
criterion = nn.CrossEntropyLoss(ignore_index=0)

print("🧠 Training started...")
for epoch in range(EPOCHS):
    model.train()
    total_loss = 0
    for src, tgt in train_loader:
        src, tgt = src.to(DEVICE), tgt.to(DEVICE)
        tgt_input = tgt[:, :-1]
        tgt_output = tgt[:, 1:]

        tgt_mask = generate_square_subsequent_mask(tgt_input.size(1)).to(DEVICE)
        src_pad_mask = create_pad_mask(src).to(DEVICE)

        logits = model(src, tgt_input, src_pad_mask, tgt_mask)
        loss = criterion(logits.reshape(-1, logits.size(-1)), tgt_output.reshape(-1))  

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    print(f"Epoch {epoch+1}/{EPOCHS} | Loss: {total_loss/len(train_loader):.4f}")

torch.save(model.state_dict(), 'transformer_translator.pth')
print(" Model saved as transformer_translator.pth")

def translate(model, sentence, eng_vocab, spa_vocab, max_len=MAX_LENGTH):
    model.eval()
    rev_vocab = {v: k for k, v in spa_vocab.items()}
    tokens = re.sub(r'[^a-z\s]', '', sentence.lower()).split()
    src_ids = [eng_vocab.get(w, eng_vocab['<unk>']) for w in tokens]
    src_tensor = torch.LongTensor([src_ids]).to(DEVICE)
    src_pad = create_pad_mask(src_tensor).to(DEVICE)

    tgt_ids = [spa_vocab['<sos>']]
    for _ in range(max_len):
        tgt_tensor = torch.LongTensor([tgt_ids]).to(DEVICE)
        tgt_mask = generate_square_subsequent_mask(len(tgt_ids)).to(DEVICE)
        with torch.no_grad():
            output = model(src_tensor, tgt_tensor, src_pad, tgt_mask)
        next_token = output[0, -1].argmax().item()
        if next_token == spa_vocab['<eos>']:
            break
        tgt_ids.append(next_token)
    return ' '.join([rev_vocab.get(i, '<unk>') for i in tgt_ids[1:]])

print(" Ready to translate. Type 'exit' to quit.")
while True:
    inp = input("EN> ")
    if inp.lower().strip() == 'exit':
        break
    result = translate(model, inp, eng_vocab, spa_vocab)
    print("ES>", result)
